### ISSUE TITLE
"Handshake failed.  Error in the certificate."

### ISSUE THREAD

**Issue by n8henrie** [GothenburgBitFactory/taskwarrior#2233]
I'm trying to get set up to use TaskWarrior using a remote server as a sync station. I don't have a domain name for the server, but it has a static IP address, and the documentation makes it seem like that should work.

I'm running Arch linux, 5.3.11. I installed `task` and `taskd` via pacman, and all my packages (including gnutls) are up to date. I don't see any active bug reports for `taskd` with arch.

I've tried following:
- https://taskwarrior.org/docs/taskserver/taskwarrior.html
- https://taskwarrior.org/docs/taskserver/troubleshooting-sync.html
- https://wiki.archlinux.org/index.php/Taskd

I can see the `taskd` is running; when I run `task sync init` from the same machine, I get:

```
Handshake failed.  Error in the certificate.
Sync failed.  Could not connect to the Taskserver.
```

I see this same error with any of the changes I've tried below.

We'll say the machine's static IP address is 1.2.3.4 and it is configured for port 12345.

In my taskrc, I have tried `taskd.server=localhost:12345` and `taskd.server=1.2.3.4:12345`. In `/var/lib/taskd/config` I have tried `server=*:12345`, `server=0.0.0.0:12345`, and `server=1.2.3.4:12345`.

I have tried deleting and regenerating all of .pem files, both for the server and the client.

I have confirmed that the CN in the certs is `1.2.3.4` (via both methods at https://gitpitch.com/GothenburgBitFactory/taskserver-troubleshooting#/1/11)

```console
$ task diag

task 2.5.1
   Platform: Linux

Compiler
    Version: 8.2.1 20180831
       Caps: +stdc +stdc_hosted +LP64 +c8 +i32 +l64 +vp64 +time_t64
 Compliance: C++11

Build Features
      Built: Nov  9 2018 23:33:36
      CMake: 3.12.4
    libuuid: libuuid + uuid_unparse_lower
  libgnutls: 3.5.19
 Build type:

Configuration
       File: /home/n8henrie/.taskrc (found), 1643 bytes, mode 100644
       Data: /home/n8henrie/.task (found), dir, mode 40755
    Locking: Enabled
         GC: Enabled
    $VISUAL: nvim
     Server: localhost:52260
         CA: /home/n8henrie/.task/ca.cert.pem, readable, 3775 bytes
      Trust: strict
Certificate: /home/n8henrie/.task/n8henrie.cert.pem, readable, 3714 bytes
        Key: /home/n8henrie/.task/n8henrie.key.pem, readable, 24745 bytes
    Ciphers: NORMAL
      Creds: henrie/n8henrie/************************************

Hooks
     System: Enabled
   Location: /home/n8henrie/.task/hooks
             (-none-)

Tests
      $TERM: screen (343x71)
       Dups: Scanned 1 tasks for duplicate UUIDs:
             No duplicates found
 Broken ref: Scanned 1 tasks for broken references:
             No broken references found
```

I have confirmed the port is open using nmap from a different computer, and I can seen taskd listening with `sudo lsof -i :12345`.

Thanks in advance for any recommendations.

**Comment by corbolais** [comment]
**TL;DR:** I guess, it boils down to: keep taskd config and task config absolutely in sync AND match cert CN, hostname/IP AND mind the backslash issue mentioned below.

Here's kinda my step-by-step how I went down the rabbit hole:

You could always listen in using `tcpdump -ni <ifname> port 12345` to see whether there's someone talking to your taskd at all. Since you get a `Handshake failed. Error in the certificate.` there is a connection.

In my experience, it is crucial for `taskd.server=`**1.2.3.4**`:12345` being identical to the PEMs CN. localhost and stuff won't work which is quite nasty since the cert can be verified w/o imposing this restriction on the `taskd.server` semantics. This identity requirement holds even for `~/.taskrc:taskd.server` (on the client running task): set to the servers IP, FQDN or whatever is appropriate for any networking setting on the client's `taskd.server` will make it fail as well **iff** it differs from the certs' CN. I cannot stress it enough: it has to be **identical** to the certs' CN. Then again, the certs' CN and the servers' IP/hostname/FPDN have also to be **identical**, whatever they're set to (would be the servers' IP in your case).

In the end, you could use `gnutls-cli 1.2.3.4:12345` to get any clue there is left regarding a cert or config mismatch. This was most helpful to me in narrowing it down to this identity enforcement in places not really requiring, and most obviously you not expecting it.

Oh, and BTW, use gnutls to create the certs. I wasn't successfull in (re-)creating them using openssl (maybe some permutation of config/cert issue got me on the way, who knows).

One other thing to watch out for is the weird notation syntax:

taskd.certificate=\/home\/me\/.task\/my.cert.pem

Omit the backslashes and it will fail, again. On a Linux system, mind you.

Full disclosure: I love TW and taskd and the cert verification. The oddities not so much.

HTH

**Comment by n8henrie** [comment]
I've now set the CN to `1.2.3.4` and both server and client configs to `1.2.3.4:12345`, ensured `hostname -f` is also `1.2.3.4`, rebooted the server, no change. 

I forgot to include relevant log output: `Error: Handshake failed. The TLS connection was non-properly terminated.`

I don't know if this output might give some clues, I initially thought `read:errno=0` meant 0 exit code and no errors, but just above that I now see a return code of 21.

```console
$ openssl s_client -CAfile .task/ca.cert.pem -host 1.2.3.4 -port 12345
Can't use SSL_get_servername
depth=0 CN = 1.2.3.4, O = n8henrie.com
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = 1.2.3.4, O = n8henrie.com
verify error:num=21:unable to verify the first certificate
verify return:1
CONNECTED(00000003)
---
Certificate chain
 0 s:CN = 1.2.3.4, O = n8henrie.com
   i:CN = 1.2.3.4, O = n8henrie.com, L = Shiprock, ST = New Mexico, C = US
---
Server certificate
-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----
subject=CN = 1.2.3.4, O = n8henrie.com

issuer=CN = 1.2.3.4, O = n8henrie.com, L = Shiprock, ST = New Mexico, C = US

---
Acceptable client certificate CA names
CN = 1.2.3.4, O = n8henrie.com, L = Shiprock, ST = New Mexico, C = US
Requested Signature Algorithms: RSA+SHA256:RSA-PSS+SHA256:RSA-PSS+SHA256:ECDSA+SHA256:Ed25519:RSA+SHA384:RSA-PSS+SHA384:RSA-PSS+SHA384:ECDSA+SHA384:RSA+SHA512:RSA-PSS+SHA512:RSA-PSS+SHA512:ECDSA+SHA512:RSA+SHA1:ECDSA+SHA1
Shared Requested Signature Algorithms: RSA+SHA256:RSA-PSS+SHA256:RSA-PSS+SHA256:ECDSA+SHA256:Ed25519:RSA+SHA384:RSA-PSS+SHA384:RSA-PSS+SHA384:ECDSA+SHA384:RSA+SHA512:RSA-PSS+SHA512:RSA-PSS+SHA512:ECDSA+SHA512
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 2419 bytes and written 403 bytes
Verification error: unable to verify the first certificate
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 4096 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 21 (unable to verify the first certificate)
---
read:errno=0
```

> Oh, and BTW, use gnutls to create the certs.

I used the built-in `generate` scripts, which seem to use gnutls behind the scenes.

> One other thing to watch out for is the weird notation syntax:

I think markdown may have eaten your backslashes; my config does have backslashes.

**Comment by corbolais** [comment]
1st thought was: what does `gnutls-cli` say?

> ```
> verify error:num=20:unable to get local issuer certificate
> ```

2nd read: did you try to fix the above issue? Now, I ignore why `openssl` is telling you that. Maybe `gnutls-cli` is more verbose or more to the point with it's message. Tried google?

**Comment by n8henrie** [comment]
Thanks for your input.

> Tried google?

Yes. Lots of Google.

> unable to get local issuer certificate

I thought that would be normal for a self-signed cert, but I guess I have no idea: https://confluence.atlassian.com/bitbucketserverkb/ssl-certificate-problem-unable-to-get-local-issuer-certificate-816521128.html

> Maybe gnutls-cli is more verbose or more to the point with it's message

I also tried:

```console
$ gnutls-cli 1.2.3.4:12345
*** Fatal error: Error in the certificate.
Processed 150 CA certificate(s).
Resolving '1.2.3.4:12345...
Connecting to '1.2.3.4:12345'...
- Successfully sent 0 certificate(s) to server.
- Server has requested a certificate.
- Certificate type: X.509
- Got a certificate list of 1 certificates.
- Certificate[0] info:
 - subject `O=n8henrie.com,CN=1.2.3.4', issuer `C=US,ST=New Mexico,L=Shiprock,O=n8henrie.com,CN=1.2.3.4', serial ..., RSA key 4096 bits, signed using RSA-SHA384, activated `2019-12-11 21:02:09 UTC', expires `2020-12-10 21:02:09 UTC', pin-sha256="..."
	Public Key ID:
		sha1:...
		sha256:...
	Public Key PIN:
		pin-sha256:...

- Status: The certificate is NOT trusted. The certificate issuer is unknown. The name in the certificate does not match the expected. 
*** PKI verification of server certificate failed...
```

**Comment by corbolais** [comment]
> ```
> - Status: The certificate is NOT trusted. The certificate issuer is unknown. The name in the certificate does not match the expected. 
> ```

So `gnutls-cli` is more to the point, I guess.

Seems like you need to get your certs sorted.

Did you succeed?

**Comment by n8henrie** [comment]
Thanks for the edit and checking on your output. Yes, I assume it's something to do with the certificate.

Does  anybody else use a bare (static) IP address for their SSL cert CN? I wonder if that might be the problem.

> Are the taskd and task binaries version-compatible?

I have no idea. They're both what's avail in the Arch repos, so I assume so: taskd 1.1.0 built for linux, task 2.5.1

**Comment by n8henrie** [comment]
I found one incorrect cert, regenerated them all again, and now OpenSSL seems satisfied: `Verify return code: 0 (ok)`

But this might be an issue:

```console
$ task diag | grep gnutls
  libgnutls: 3.5.19
$ taskd diag | grep gnutls
   libgnutls: 3.6.9
```

**Comment by corbolais** [comment]
> Does anybody else use a bare (static) IP address for their SSL cert CN? I wonder if that might be the problem.

Nope, but that alone shouldn't be a problem. Use `/etc/hosts` and set the name there, on both hosts. Then recreate certs etc. and install them. Works fine here!

**Comment by corbolais** [comment]
Good catch with the gnutls-lib version (mis-)matching. That was in fact an issue here, once.

As long as `taskd` is complaining about the connection and/or cert issues, I'd continue digging that rabbit hole. There seems still something amiss. Maybe, try again using `gnutls-cli`, telling you the same (wondering, b/c openssl seems happy)? 

I'm curious about what was the issue. Once you got it resolved, that is. ;-)

**Comment by n8henrie** [comment]
Could you show me your `/etc/hosts`? Perhaps that is wrong.

The only thing in mine is:

```plaintext
1.2.3.4 1.2.3.4 myRemoteAlias
```

**Comment by n8henrie** [comment]
Well, I set up another duckdns subdomian, and using that as the CN instantly got everythign working.

Removed everything from `/etc/hosts`, using the Arch-repo provided binaries (gnutls mismatch and all), instantly everything just worked. Even changed the taskd setting back to `server=*:12345` and that still works (though breaks if I change `taskd.server` in `.taskrc`).

¯\\(ツ)/¯

If anyone figures out how to get it working with a bare IP address I'd be interested in figuring out what I was doing that wasted the last day effort, but for now it's working with duckdns so I'll close.

**Comment by corbolais** [comment]
> 1.2.3.4 1.2.3.4 myRemoteAlias

Try with `1.2.3.4 myRemoteAlias`

**Comment by corbolais** [comment]
Oh, you got it working! Well done!

And it was the hostname thing after all. Bummer. So, maybe, I did that `/etc/hosts` thing exactly for this reason and didn't take a note about it.

Well IIRC, you had some invalid certs in between, too.

Glad you got it working.
